### PR TITLE
Add analyzer for logic inversion

### LINKED ISSUE #222: [Analyzer] Logic Inversion
Simple Task: We need an expression analyzer for BooleanNot_ Node.
We want to check for this:
```php
if (!(a == b)) // use != ... same check for <,>,<=,>=, ===, !=, !==
```

How to do it:
1. Create Analyzer (you can use this one as a template: https://github.com/ovr/phpsa/blob/master/src/Analyzer/Pass/Expression/ErrorSuppression.php )
and register it to Expr\BooleanNot_. Now check that the expr property is an instanceof one of the operators listed above. You can find all the operator nodes here:
https://github.com/nikic/PHP-Parser/tree/master/lib/PhpParser/Node/Expr/BinaryOp

2. Add it to PHPSA\Analyzer\Factory (registerExpressionPasses)

3. Write Tests (analyze-fixtures directory): One example without logic inversion and at least one with. (Best would be an example for every operator probably) (again look at the existing ones for expressions like ErrorSuppression)

Check our Contribution Guide and create a pull request.
Thanks :cake:

### PR DESCRIPTION
**Pull request by Lexty**
Hey!

Type: new feature

Link to issue: fix #222 

This pull request affects the following components **(please check boxes):**

* [ ] Core
* [x] Analyzer
* [ ] Compiler
* [ ] Control Flow Graph
* [ ] Documentation

**In raising this pull request, I confirm the following (please check boxes):**

- [x] I have read and understood the [Contributing Guidelines](https://github.com/ovr/phpsa/blob/master/.github/CONTRIBUTING.md).
- [x] I have checked that another pull request for this purpose does not exist.
- [x] I wrote some tests for this PR.

Small description of change:
Adds a new Analyzer that raises a notice for expressions: `!(a == b)`, `!(a != b)`, `!(a === b)`, `!(a !== b)`, `!(a < b)`, `!(a <= b)`, `!(a > b)`, `!(a >= b)`, `!(!a)`.
Note: **Also adds not described in #222 expression `!(!a)`.**

Thanks

### CODE REVIEW

File: src/Analyzer/Pass/Expression/LogicInversion.php [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+ */
+namespace PHPSA\Analyzer\Pass\Expression;
+
+use PhpParser\Node\Expr\BooleanNot;
+use PHPSA\Analyzer\Pass\AnalyzerPassInterface;
+use PHPSA\Context;
+
+class LogicInversion implements AnalyzerPassInterface
+{
+    protected static $map = [
```
Comment (ddmler) [inline src/Analyzer/Pass/Expression/LogicInversion.php]: make this not static

File: src/Analyzer/Pass/Expression/LogicInversion.php [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    ];
+
+    /**
+     * @param BooleanNot $expr
+     * @param Context $context
+     * @return bool
+     */
+    public function pass(BooleanNot $expr, Context $context)
+    {
+        if ($expr->expr->getType() === 'Expr_BooleanNot') {
```
Comment (ddmler) [inline src/Analyzer/Pass/Expression/LogicInversion.php]: this is not needed we have another analyzer for it (called MultipleUnaryOperators)

File: src/Analyzer/Pass/Expression/LogicInversion.php [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+            // BooleanNot expression needs a different message.
+            $msg = 'Use "a" expression instead of "!(!a)".';
+        } else if (!array_key_exists($expr->expr->getType(), static::$map)) {
```
Comment (ddmler) [inline src/Analyzer/Pass/Expression/LogicInversion.php]: use $this->map when you removed the static.

File: tests/analyze-fixtures/Expression/LogicInversion.php [hunk 4]
```diff
@@ -0,0 +1,169 @@
+<?php
+
+/**
+ * @author Medvedev Alexandr https://github.com/lexty <alexandr.mdr@gmail.com>
+ */
+namespace Tests\Analyze\Fixtures\Expression;
+
+class LogicInversion {
```
Comment (ddmler) [inline tests/analyze-fixtures/Expression/LogicInversion.php]: the `{` has to be on the next line

**Comment by Lexty** [comment]
@ddmler Update commit according to your comments.

**Comment by ddmler** [comment]
no problem, thanks @Lexty :+1:

**Comment by ovr** [comment]
Rly cool feature 👍 Thanks @Lexty for PR and @ddmler for Review
Great job 🍻

**Pull request merged**